RFC0759 - React Native Frameworks (#759)
react-native-community/discussions-and-proposals: proposals/0759-react-native-frameworks.md

Over the years, the ecosystem has developed support for other platforms which we call [Out-of-Tree platforms (OOT)](https://reactnative.dev/docs/out-of-tree-platforms).

Each OOT platform comes with its own documentation, setup instructions, and tools/libraries to develop applications on the new target platform.

We consider the Framework and Platform concepts as **orthogonal**, i.e. you can use a Framework to ship to multiple platforms, and a Platform can be targeted by multiple Frameworks. In this context the role of a Framework is to abstract complexity and make it easier to ship to multiple Platforms providing a cohesive Developer Experience (see below).

## Responsibilities

This section describes where we draw the line between the React Native package and Frameworks.

> [!NOTE]
> For the sake of simplicity, we will be focusing on the React Native package (i.e. the NPM package called `react-native` which offers first-party support for Android & iOS).

### What does the React Native package do

> The React Native package is responsible for **core APIs and capabilities** to build native apps. React Native is also responsible for offering **APIs** that allow Frameworks to deliver their experience. When such APIs are missing, it’s our interest to create them and have a conversation with Framework developers to inform their development, and help framework developers merge changes back into React Native.

Reviewer: i believe the idea is that in a future RFC, or series of RFCs, the plan is to go into more detail on what the exact core APIs and capabilities will be. that may be worth calling out here, so folks know that this information is coming

Author: Yup added in 2387901

  - The Native Modules API
  - APIs of Core Native Modules provided by React Native
- **Runtime & Initialization** - The React Native package is responsible for offering an API to get access to the runtime and defining its lifecycle. The bridge and the bridgeless API are part of this API surface.
- **Internals mechanism** - The React Native package is responsible of encapsulating away complexity and handling mechanisms such as:
  - The threading model
  - Interoperability layers between different architectures
- **JS Engine Integration** - The React Native package is responsible for offering an API to integrate a target JS engine. React Native is distributed with built-in integrations for Hermes and JavaScriptCore. Third-party engines (like V8) may be integrated but are not directly supported.
- **Bundler integration** - The React Native package defines the integration with Metro, the React Native bundler. Part of this API surface are:
  - Building the JS and asset bundle
  - The development server
  - Debugging integrations
  - Sourcemap support

>   - Error and Crash reporting

Reviewer: we've discussed this before but this is definitely one of the areas that we'd love to be an extension point, so there is a public api that allows us swap out the redbox and other 'dev support' UI with alternatives (similar to how react-dom oriented frameworks provide their own error handling experiences that can be more framework-context-aware)

also related to this is the mechanism for loading a bundle. we often get bit by changes here and they are not the easiest to test for in CI, for example in react-native 0.73:
- https://github.com/expo/expo/pull/26630
- https://github.com/expo/expo/pull/26441

Author: > one of the areas that we'd love to be an extension point, so there is a public api that allows us swap out the redbox and other 'dev support' UI with alternatives

Yup I agree here.

I'd like to have @motiz88 @huntie @RobHogan to help with the framing as this falls under their domain.

Here is a (non-exhaustive) list of areas of what the React Native package is **not directly focusing** on, and is delegating this responsibility to frameworks:

- **Project Scaffolding** - Frameworks can offer tools and templates to create new projects so that users can get started developing in a matter of seconds.
- **Tooling for React Native upgrade DevX** - Similarly to project scaffolding, framework can offer tools to make sure the project can easily keep up with the latest React Native version. Tools such as the upgrade helper, Expo Config Plugins and others fit in this category.
- **Navigation and Routing solutions** - React Native doesn’t offer a navigation or a routing solution out of the box. Frameworks can either offer their own solution or recommend a preferred navigation solution to use.
- **Managing the build experience** - Framework can offer tools to simplify the native builds and move to fully managed native builds that are happening in the cloud.
- **Manage the publishing experience** - Frameworks can offer solutions to publish applications to public stores (App Store/Play Store and others) either locally or managed in the cloud.

> - **Providing OTA updates** - Frameworks can offer mechanisms to deploy Over-the-air updates to users. Tools such as EAS Update & Codepush fall into this category.

Reviewer 2: Maybe the framework should provide an interface for any third party to develop their own OTA update functionality, so developer could choose whatever provider they want but to ensure the basic OTA functionality work the same across all different providers

Author: As @brentvatne mentioned in [this other comment](https://github.com/react-native-community/discussions-and-proposals/pull/759#discussion_r1476519803), yes we'd love to have more formal APIs and RFCs a around those APIs.

Historically, we never really provided clear boundaries/APIs for framework developers, OTA updates is one of this scenario but not the only one.
As the React Native matures, we'll be providing more precise API boundaries and documentation for developers.

> so developer could choose whatever provider they want but to ensure the basic OTA functionality work the same across all different providers

Yup, React Native will be providing the core functionalities, and Frameworks can customize them by offering top-notch developer experience with their own flavor.

- **Auto-Linking libraries** - Frameworks can offer mechanisms to simplify linking of external 3rd party libraries.
- **Support new targets via Out-of-tree Platforms** - Frameworks can offer support for additional targets by using OOT Platforms, in order to offer a more cohesive Development experience.

## What do we recommend to React Native app developers?

> As mentioned before, we believe React Native Framework offers all the tools you’ll eventually need to get to production quickly while offering a superior developer experience. As we mentioned before:

Reviewer 3: "we believe [a] React Native Framework"?

> You’re either using a React Native Framework or you’re building your own React Native Framework

> We believe that an existing React Native Framework should be the default solution for the **majority of app developers**, specifically for beginners and developers which lack a native background.

Reviewer 3: ```suggestion
We believe that an existing React Native Framework should be the default solution for the **majority of app developers**, including developers who have more of a web background than a native background.
```
I find this wording to be more accurate and welcoming, in a way. General software engineering experience and native development skills both are valuable with a Framework, and I suspect a good part of the audience has a web background, coming from React DOM.

- PRs and Issues coming from the Framework developers will receive higher priorities.

### Requirements

Here is a short summary of what are the requirements for React Native Frameworks.

#### DOs of Frameworks

- **Focus on developer & user experience** - We believe Frameworks should aim to provide a top-notch developer experience. Frameworks that fail to provide good documentation or developer workflows are not likely high quality frameworks, and our users will most likely not want to use them. Similarly, we want to promote frameworks that provide a top-notch user experience and do not ignore basic user needs, such as accessibility, performance, and delightful UIs.
- **Focus on building cohesive experiences** - Users are looking into deploying universal apps across multiple targets. We’re delegating this responsibility to frameworks that can provide better experiences to users and craft cohesive developer experiences between mobile, web, desktop and beyond.
- **Focus on abstracting complexity away** - Any initiative that simplifies the developer workflow is welcomed in the framework domain.

> - **Be popular** - We’ll only be recommending solutions that are widely known and adopted by the community.

Reviewer 4: How do we quantify that? (stars, npm downloads, or companies using it in production) Who's "we" in this context? I don't think this is something we have to answer right now, but I believe this can become more important as new frameworks are under way.

- **Open source** - Recommended frameworks must be free, open source, projects. We allow for companies to provide paid services on top of the Open Source framework, but these services should not be a requirement to use the framework and should not have vendor lock-in.

#### DON’Ts of Frameworks

- **Do not hard fork react-native** - As a rule of thumb, you should not GitHub-fork react-native and re-publish it. If you’re doing so, it’s probably due to a lack of APIs that React Native should provide. Please reach out to us and attempt to de-fork your framework.

> - **Do not soft fork the core APIs** - We want to make sure developers need to learn only once how to write a native component. Building abstraction and wrappers is accepted, forking is not, as it makes it harder for users to share components/modules between targets.

Reviewer 5: In those terms, is it acceptable to remove certain parts of `react-native`? For example, I create a React Native framework that removes some props I consider harmful in the `StatusBar` module, or entirely removes certain APIs / Components, such as dropping `Modal` to provide a better alternative.

Author: > In those terms, is it acceptable to remove certain parts of react-native? 

As a principe: nope it's not. The reality is a bit different in the sense that we know there are some core components/API with better community alternatives.

Our goal is to work with Framework/Library developers to understand those gaps and streamline the core as much as possible. This follows the same direction of the [Lean Core initiative](https://github.com/facebook/react-native/issues/23313) in 2019. We want the core API to be streamlined and extensible so you don't need to "remove" anything from react-native.

> - **Do not soft fork the core APIs** - We want to make sure developers need to learn only once how to write a native component. Building abstraction and wrappers is accepted, forking is not, as it makes it harder for users to share components/modules between targets.

Reviewer 6: This bullet is not totally clear to me. Why is "learn only once how to write a native component" the particular example used here? I'm not sure how to connect it with the intention of the rule.

Also, I'm not sure hard-fork and soft-fork are widely-known terms, I would consider defining those more clearly.

Author: Does my other comments clarify here? https://github.com/react-native-community/discussions-and-proposals/pull/759#discussion_r1478854364

> - **Do not fork the ecosystem** - It’s crucial that `react-native-*` libraries can be easily shared and ported to other targets or frameworks (with little to no work). Frameworks should aim at building libraries that work across the whole ecosystem and not solutions that silo the users away.

Reviewer 6: Again, I'm having trouble understanding why "forking" is the focus here, especially for a library. Like if I'm using the third-party `react-native-foo` library, I need to add some functionality but the maintainer is not responsive, I would expect I could fork and add what I'd like to my version. What's the problem there?

Author: > Like if I'm using the third-party react-native-foo library, I need to add some functionality but the maintainer is not responsive, I would expect I could fork and add what I'd like to my version. What's the problem there?

Forking an unmaintaied library is actually encouraged, as you're providing a solution to one of the biggest pain point of React Native ([source](https://results.stateofreactnative.com/opinions/opinions_pain_points_multiple)).

The sentence here refers to "forking the ecosystem". We want to avoid situations where a framework `foo-` forks a core API and starts encouraging developers do develop libraries on top of that forked API. So you'll end up having say a `foo-push-notification` libraries, rather than a `react-native-push-notification` camera that can easily be used between different frameworks.

> - **Do not redistribute prebuilts** - React Native ships native prebuilts for several targets. Do not attempt to republish them as this reduces security for users and is essentially yet another fork.

Reviewer 3: Relatedly, I believe the static library for Hermes will need to be signed by Meta for the iOS in the future to secure part of the SBOM (app review policy). Probably too tangential of a detail to mention in the RFC, but it is another reason that reinforces this point here.

Reviewer 7: Hi @ide, sorry for the late reply. For the privacy updates Apple is pushing, we have an internal effort to fix them. 
For most of them, we will include a privacy manifest plist in the template, which Frameworks could take as starting point if they use other restricted APIs.

For `hermes` specifically, we are talking with Apple directly to understand what to do. We might end up removing/replacing the offending methods to avoid to sign the binary.

#### Note for Platforms developers

The aforementioned Do/Dont’s don’t apply strictly to **Platforms developers**. While they do represent general good practices in the software development field, targeting a new platform poses a series of **unique challenges**.

Specifically, when supporting platforms, having a fork might actually be the preferred solution, based on the amount of custom logic that is necessary to maintain. Similarly, certain platforms require redistributing prebuilts for new binary architectures which are not already provided by React Native.

The React Native team is committed to fostering innovation and supporting new platforms by offering APIs and artifacts for those new platforms, whenever possible.

## Evolving the React Native Community Template & CLI

Users who have been using the React Native **Community CLI & Template**, might be wondering how React Native Frameworks affect them.

> We want to reassure our users that are using those tools and clarify that those tools will **still be supported** **and updated**, and you can keep on using them if you wish. However, we’ll be recommending new users and **newly created mobile apps** to use a React Native Framework from now on. Similarly, you’re free to migrate to a React Native Framework if you intend to.

Reviewer: super important point here for many folks who will come across the rfc! this rfc has an absolutely minimal impact on developers that want to use `@react-native-community/cli` tooling - effectively, when the changes roll out:

- initialize new apps with `npx @react-native-community/cli init` instead of `npx react-native init` if you prefer community tooling
- `npx react-native config` becomes `npx @react-native-community/cli config` and `npx react-native doctor` becomes something like `npx react-native-doctor`. commands like `run-ios` and `start` are considered to be part of `react-native` itself, as discussed below

Similar to the Template, we envision having a **private CLI** inside the react-native repository to support the internal use cases mentioned above. This will be achieved by evolving each command of the CLI as follows:

##### Commands to be moved inside react-native core

Commands that live inside `react-native` should be extensible enough to be **used by Frameworks or libraries, **if they wish to**.**

- `clean, build-android, build-ios, run-android, run-ios, log-ios, log-android`
  - Those are wrappers around the native build system and need to be moved inside the react-native repository.
- `start, bundle`
  - Those are wrappers around Metro and [are effectively moved](https://github.com/facebook/react-native/pull/38795) inside the `react-native` repo already.
- `profile-hermes`
  - Should be moved inside react-native as part of default debugging/profiling experience with Hermes

##### Commands to be removed

- `upgrade`
  - This command has been removed and now only points to the upgrade helper
  - Potentially can be removed entirely
- `ram-bundle`
  - Can be removed as not relevant anymore

##### Commands to be left inside @react-native-community/cli

> - `config` and Auto Linking of dependencies.

Reviewer 4: My only worry is the potential future fragmentation in available module structures and their support by frameworks. Let's imagine we have 4 frameworks in about 2 years from now. If there's no standard way to link native dependencies, it may impose additional maintenance burden on library authors to support all popular architectures and cause confusion amongst developers when installing a library. 

Theoretically, there isn't much that can go wrong here. As long as we're using CocoaPods and Gradle under the hood, all solutions would be similar as there's a limited number of scenarios to support.

I wonder whether Community CLI and some of its internals could be rebranded as "Bare React Native Framework" package that other framework developers can use to build on top. It could provide some essential requirements as to what workflows have to be supported in order to pass the compliance test. It could also be one of the requirements when endorsing a framework that it must support certain basic workflows, such as auto-linking dependencies.

Reviewer 5: I agree on this one.

Let's say tomorrow there's a new framework that chooses to have a custom CLI and autolinking system because they truly think it's the best way to improve their users' DX. This decision will create additional work for library maintainers, as they will have to develop new, specific plugins.

That's already (kinda) an issue with Expo, as some libraries require editing additional "native" files. Expo has its custom API for that (using `@expo/config-plugins`), which actually helps users a lot. However, it's not standardized. Imagine four or five frameworks each having their custom way to create these kinds of plugins.

Reviewer: > Let's imagine we have 4 frameworks in about 2 years from now. If there's no standard way to link native dependencies, it may impose additional maintenance burden on library authors to support all popular architectures and cause confusion amongst developers when installing a library.

I think there could potentially be room for something like autolinking in core, but I'm hesitant to suggest that we should use what we have today, because it is rather slow and limited (eg: it's the main speed bottleneck for us in [@expo/fingerprint](https://github.com/expo/expo/tree/main/packages/@expo/fingerprint) and does not support transitive dependency linking). Another possible approach is that React Native doesn't end up providing an implementation of autolinking directly, but instead defines a common schema that library authors can use and frameworks can build support for.

> That's already (kinda) an issue with Expo, as some libraries require editing additional "native" files. Expo has its custom API for that (using @expo/config-plugins), which actually helps users a lot. However, it's not standardized. Imagine four or five frameworks each having their custom way to create these kinds of plugins.

If another framework is interested in following the same sort of pattern that Expo established with [CNG](https://docs.expo.dev/workflow/continuous-native-generation/) then I know that we at Expo would be happy to work with them to see where we should make parts of this toolchain more generic to minimize the maintenance costs for library authors. We want to and are incentivized to ensure that support for CNG is as easy as possible for library authors, because this provides a better experience for all users of the feature.

Additionally, it's worth noting that it's also possible to use `@expo/config-plugins` in other frameworks, see the usage in [react-native-test-app](https://github.com/microsoft/react-native-test-app) for reference, or [byCedric/custom-prebuild-example](https://github.com/byCedric/custom-prebuild-example) which demonstrates using config plugins to generate macOS and Windows projects.

@blakef summarized where we landed on this topic in a previous RFC: https://github.com/react-native-community/discussions-and-proposals/pull/637#issuecomment-1559339292

Reviewer 5: This could be included in the `@react-native-community/cli-tools` package and benefits the whole ecosystem right now.

I already do similar stuff in the [react-native-bootsplash CLI](https://github.com/zoontek/react-native-bootsplash/blob/master/react-native.config.js), would love to have this available and white label, as it's currently kinda hard for me to write a plugin for expo without duplicating the whole logic.

Imagine being able to edit any `app.json` like this:

```json
{
  "name": "RNBootSplashExample",
  "displayName": "RNBootSplashExample",

  "plugins": [
    [
      "react-native-bootsplash",
      {
        "background": "#F5FCFF",
        "logo": "./svgs/light_logo.svg",
        "logoWidth": 100
      }
    ]
  ]
}
```

Author: > Another possible approach for core is that React Native doesn't end up providing an implementation of autolinking directly, but instead defines a common schema that library authors can use and frameworks can build support for.

That's the direction where we're heading towards.
Practically autolinking as it is today, it's composed by two components:

1. Integration points - They should stay inside `react-native`. Those are APIs like `ReactPackage`. Practically speaking that's where you can 'feed' modules/components/libraries to core so that they're accounted as part of the build and runtime of react native.
2. Discovery logic - This should stay outside `react-native`. This is where `react-native config` and other discovery mechanism are executed (i.e. on Android we search for `.java`/`.kt` files which implements the `ReactPackage` API exposed by core). Here frameworks can implement their own discovery mechanisms.

        - [Commands to be extracted as standalone commands](#commands-to-be-extracted-as-standalone-commands)
    - [The init command](#the-init-command)

## What is a React Native Framework?

React Native allows developers to use the [React programming paradigm](https://react.dev/learn/thinking-in-react) to ship applications to native targets. The React Native team is focused on creating **core APIs** and **functionalities** that fit the most general use case when developing native apps (not just restricted to Android & iOS).

Shipping native apps to production usually requires a set of tools & libraries that are not provided by default as part of React Native, but are still crucial to hit the ship to production. Examples of such tools are: support for publishing applications to a dedicated store, or support for routing and navigation mechanisms.  

When those tools & libraries are collected to form a cohesive framework built on top of React Native, we call this a **React Native Framework**.

An example of Open Source React Native Framework is [Expo](https://expo.dev/).

> If you’re not using a React Native Framework, but you’re building production-ready applications, you probably have a collection of scripts and automations that helps you develop and ship applications to production. This collection of infrastructure is a bespoke React Native Framework that you built around your own needs and processes. For this reason, is easier to think about React Native Framework as:

Reviewer 8: Avoiding "but" to simplify the message
```suggestion
If you’re building production-ready applications without using React Native Framework, you probably have a collection of scripts and automations that helps you develop and ship applications to production. This collection of infrastructure is a bespoke React Native Framework that you built around your own needs and processes. For this reason, is easier to think about React Native Framework as:
```

A practical example of bespoke React Native Framework is how Meta uses React Native internally. We do have a collection of libraries, scripts and tools that make the integration of React Native easier with the various apps where it's used.

For the rest of this document we will clearly state what are the responsibilities of React Native Frameworks, so that if you don’t want to use an available solution, you know what are the boundaries of React Native.

> [!NOTE]
> We do acknowledge that “Framework” is an overloaded term here, as React Native itself is a UI framework which uses React. For the sake of simplicity in the rest of this document, the “Framework” term will always refer to a React Native Framework — a set of tools & libraries built on top of React Native.

## Why are we defining this now?

You might be wondering why we are defining those concepts now. React Native has been **growing in popularity, scope & maturity** over the recent years. This is mostly thanks to one of the strongest components of React Native: our **ecosystem**.

The Authors & Maintainers of React Native libraries and tools have played such a crucial role in the development of React Native.

> The intention of this document is **not** to drop responsibilities on their shoulders. The React Native team is fully committed to supporting React Native in the same capacity as we’ve been doing to this day. But we believe that by clearly defining our **boundaries**, we could build a **better ecosystem** for everyone and foster further growth to increase adoption and stability of React Native.

Reviewer 8: ```suggestion
The intention of this document is **not** to drop responsibilities on their shoulders. The React Native team is fully committed to supporting React Native in the same capacity as we’ve been doing to this day. We believe that by clearly defining our **boundaries**, we could build a **better ecosystem** for everyone and foster further growth to increase adoption and stability of React Native.
```

- **Tooling for React Native upgrade DevX** - Similarly to project scaffolding, framework can offer tools to make sure the project can easily keep up with the latest React Native version. Tools such as the upgrade helper, Expo Config Plugins and others fit in this category.
- **Navigation and Routing solutions** - React Native doesn’t offer a navigation or a routing solution out of the box. Frameworks can either offer their own solution or recommend a preferred navigation solution to use.
- **Managing the build experience** - Framework can offer tools to simplify the native builds and move to fully managed native builds that are happening in the cloud.
- **Manage the publishing experience** - Frameworks can offer solutions to publish applications to public stores (App Store/Play Store and others) either locally or managed in the cloud.
- **Providing OTA updates** - Frameworks can offer mechanisms to deploy Over-the-air updates to app users. Tools such as EAS Update & Codepush fall into this category.
- **Auto-Linking libraries** - Frameworks can offer mechanisms to simplify linking of external 3rd party libraries.

> - **Support new targets via Out-of-tree Platforms** - Frameworks can offer support for additional targets by using OOT Platforms, in order to offer a more cohesive Development experience.

Reviewer 8: ```suggestion
- **Support new targets via Out-of-tree Platforms** - Frameworks can offer support for additional targets by using OOT Platforms, in order to offer a more cohesive development experience.
```

#### DON’Ts of Frameworks

- **Do not hard fork react-native** - As a rule of thumb, you should not GitHub-fork react-native and re-publish it. If you’re doing so, it’s probably due to a lack of APIs that React Native should provide. Please reach out to us and attempt to de-fork your framework.
- **Do not soft fork the core APIs** - We want to make sure developers need to learn only once how to write a native component. Building abstraction and wrappers is accepted, forking is not, as it makes it harder for developers to share components/modules between targets.
- **Do not fork the ecosystem** - It’s crucial that `react-native-*` libraries can be easily shared and ported to other targets or frameworks (with little to no work). Frameworks should aim at building libraries that work across the whole ecosystem and not solutions that silo the developers away.
- **Do not redistribute prebuilts** - React Native ships native prebuilts for several targets. Do not attempt to republish them as this reduces security for app users and is essentially yet another fork.

> - **Do not depend on the internals** - Do not depend on internal APIs. As a rule of thumb, depending on private/internal API (i.e. those from the `internal` package or with `unstable_` prefix) is not recommended and will cause breakages in the future. Specifically, we might consider rotating private APIs (changing the name) in the future to discourage its usages. When there is a need for Frameworks to access those internal APIs, we’ll be annotating them to make them easy to recognize it (i.e. with the `@FrameworkAPI` annotation).

Reviewer 8: Is annotating private APIs for frameworks a temporary step before making it public?

Author: Nope not really. APIs that are Framework annotated are those where we reached an agreement with Framework developers that they'll be the sole users. So they should not be considered stable as they could break if we iterate further on them.

To fully understand the role of the template, it’s crucial we understand what are the use cases of the template as of today.

The React Native template is used for:

1. **Internal** use cases (Meta engineers/Contributors)
    1. The React Native CI is using it to test that a new project can be successfully created on every PR/Commit ([example job definition here](https://github.com/facebook/react-native/blob/28dfdb22ebaff7055fd8bb4b8847e018181f849d/.circleci/config.yml#L769))
    2. The Release Crew is using it to test that a new app can be created, built and executed correctly **before releasing** a new version of React Native ([docs on this process are here](https://reactnative.dev/contributing/release-testing))
    3. Issues and bug reports require a **reproducer**, being either an Expo Snack or an instance of a template project ([see here for the reproducer](https://github.com/react-native-community/reproducer-react-native)).
2. **External** use cases
    1. React Native developers that are **not using a Framework** because they haven’t migrated to it yet

>     2. React Native developers that are **not using a Framework** such as Expo because it doesn’t satisfy their requirements

Reviewer 8: Is it worth to directly reference the need for using out-of-tree platforms not supported by Expo (because some are supported)?

Author: Yup I've added a sentence about it here b5f463f

    6. New React Native developers that are creating a new app. They might either not know how to use a Framework, or they might seek the official/bare/**Meta-recommended** way to use React Native.

>     7. Developers looking into integrating React Native into frameworks or existing apps. They can use the template as a **reference** for their own integration, or to understand how the internals work.

Author: Thank you, fixed in [f3598af](https://github.com/react-native-community/discussions-and-proposals/pull/759/commits/f3598af7f04a9f300d5a6875f49671afb4563f66)

  - Similarly, we’ll aim at **reducing the breaking changes** as much as possible.
- Update all the various community tools to refer to the new template location (e.g. the Upgrade Helper)
- Work with existing React Native Framework to identify use cases that are not properly covered by their framework

#### Community CLI

The React Native Community CLI is tightly coupled to the template. Apps like RN-Tester (and the proposed HelloWorld app) are using the community CLI to build and being tested on debug/release on CI. Similarly the community CLI is used to do release testing.

On top of this, having the community CLI living in a separate repository (`@react-native-community/cli`) introduces a number of **circular dependencies** and friction as part of the release process.

Similar to the Template, we envision having a **private CLI** inside the react-native repository to support the internal use cases mentioned above. This will be achieved by evolving each command of the CLI as follows:

##### Commands to be moved inside react-native core

Commands that live inside `react-native` should be extensible enough to be **used by Frameworks or libraries, **if they wish to**.**

> - `clean, build-android, build-ios, run-android, run-ios, log-ios, log-android`

Reviewer 9: Does that mean that auto linking currently doesn't rely on some specific behavior of the build command? Or is there some additional work required to decouple that? From what I understand, community CLI supports different build configurations including support for out-of-tree platforms (i.e. tvOS).

What I think is the plan, but it isn't clearly stated here, is for `react-native` command to be owned/maintained/published by React Native core team. If this is the case, when you later mention the command should be removed, do you only indent for the implementation to be removed from the core repo, or also making the whole option unavailable when using `react-native` command and move it under a separate community cli command.

Reviewer 8: > Does that mean that auto linking currently doesn't rely on some specific behavior of the build command? Or is there some additional work required to decouple that? From what I understand, community CLI supports different build configurations including support for out-of-tree platforms (i.e. tvOS).

There's some small refactoring work expected before moving the commands implementations to the core, because currently we're passing the config directly.

Author: > There's some small refactoring work expected before moving the commands implementations to the core, because currently we're passing the config directly.

Yup basically this. Currently autolinking relies on invoking `react-native config` so that dependencies & project configs are loaded correctly.
We'll have to decouple it so that autolinking just expects an input that framework developers can provide.

> is for react-native command to be owned/maintained/published by React Native core team

That's correct. That's already the case today by the way.

> when you later mention the command should be removed

Are you referring to the `upgrade` & `ram-bundle` command mentioned in the `##### Commands to be removed` section? As for those, we intend the to remove from the CLI entirely so it won't be possible to invoke them at all anymore. Is this consistent with your mental model @thymikee ?

- Refer to the documentation for information about alternative tools: https://reactnative.dev/docs/getting-started

Running: npx @react-native-community/cli init 

... logs as usual
```

- **react-native@>=0.76**: Inform the developer that the command has been deprecated and just fire a warning for alternatives

```bash
$ npx react-native@latest init 

⚠️ The `init` command is deprecated. 
The behavior will be changed on January 1, 2025 (344 days). 

- Switch to npx @react-native-community/cli init for the identical behavior.
- Refer to the documentation for information about alternative tools: https://reactnative.dev/docs/getting-started
```

Similarly, the `create-react-native-app` command (currently owned by Expo) will be updated to just print a warning to developers by React Native **0.75**:

```bash
$ npx create-react-native-app@latest

Need to install the following packages:
create-react-native-app@2.0.0
Ok to proceed? (y)

⚠️ This tool does not initialize new React Native projects.

It's recommended to use a framework to build apps with React Native, for example:

Expo
     npx create-expo-app@latest

React Native Community template

>      npx @react-native-community/cli init

Author: yup that could be an option
